stdlib/public/Concurrency/VoucherSupport.h

```diff
@@ -0,0 +1,101 @@
+//===--- VoucherSupport.h - Support code for OS vouchers -----------*- C++ -*-//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Support code for interfacing with OS voucher calls.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_CONCURRENCY_VOUCHERSUPPORT_H
+#define SWIFT_CONCURRENCY_VOUCHERSUPPORT_H
+
+#include "llvm/ADT/Optional.h"
+#include "swift/ABI/Task.h"
+#include "swift/Runtime/VoucherShims.h"
+#include "TaskPrivate.h"
+#include <execinfo.h>
+#include <dlfcn.h>
+
+namespace swift {
+
+/// A class which manages voucher adoption for Job and Task objects.
+class VoucherManager {
+  /// The original voucher that was set on the thread before Swift started
+  /// doing async work. This must be restored on the thread after we finish
+  /// async work.
+  llvm::Optional<voucher_t> OriginalVoucher;
+
+public:
+  VoucherManager() {
+    SWIFT_TASK_DEBUG_LOG("[%p] Constructing VoucherManager", this);
+  }
+
+  /// Clean up after completing async work, restoring the original voucher on
+  /// the current thread if necessary. This MUST be called before the
+  /// VoucherManager object is destroyed. It may also be called in other
+  /// places to restore the original voucher and reset the VoucherManager.
+  void leave() {
+    if (OriginalVoucher) {
+      SWIFT_TASK_DEBUG_LOG("[%p] Restoring original voucher %p", this, *OriginalVoucher);
+      if (voucher_needs_adopt(*OriginalVoucher)) {
+        auto previous = voucher_adopt(*OriginalVoucher);
+        swift_voucher_release(previous);
+      } else {
+        swift_voucher_release(*OriginalVoucher);
+      }
+      OriginalVoucher = llvm::None;
+    } else
+      SWIFT_TASK_DEBUG_LOG("[%p] Leaving empty VoucherManager", this);
+  }
+
+  ~VoucherManager() {
+    assert(!OriginalVoucher);
+  }
+
+  /// Set up for a new Job by adopting its voucher on the current thread. This
+  /// takes over ownership of the voucher from the Job object. For plain Jobs,
+  /// this is permanent. For Tasks, the voucher must be restored using
+  /// restoreVoucher if the task suspends.
+  void swapJob(Job *job) {
```

Comment:
Maybe `swapToJob()` rather than `swapJob()`? I was quite tempted by `adoptVoucher()` too.

Reply:
`swapToJob` seems good. The "voucher" part will be implied by the object it's called on.

Reply:
Of course I say "voucher" in a method name right after this....